ragstack/colbertbase/colbertbase/astra_db.py

```diff
@@ -0,0 +1,177 @@
+from typing import List
+from cassandra.cluster import Cluster
+from cassandra.auth import PlainTextAuthProvider
+from cassandra import InvalidRequest
+from cassandra.concurrent import execute_concurrent_with_args
+
+from .token_embedding import PassageEmbeddings
+
+def required_cred(cred: str):
+    if cred is None or cred == "":
+        raise ValueError("Please provide credentials")
+
+class AstraDB:
+    def __init__(
+            self,
+            secure_connect_bundle: str="",
+            astra_token: str=None,
+            keyspace: str="colbert128",
+            verbose: bool=False,
+            timeout: int=60,
+            **kwargs,
+    ):
+        required_cred(secure_connect_bundle)
+        required_cred(astra_token)
+
+        # self.cluster = Cluster(**kwargs)
+        self.cluster = Cluster(
+            cloud={
+                'secure_connect_bundle': secure_connect_bundle
+            },
+            auth_provider=PlainTextAuthProvider(
+                'token',
+                astra_token
+            )
+        )
+        self.keyspace = keyspace
+        self.session = self.cluster.connect()
+        self.session.default_timeout = timeout
+        self.verbose = verbose
+
+        print(f"set up keyspace {keyspace}, tables and indexes...")
+
+        if keyspace not in self.cluster.metadata.keyspaces.keys():
+            raise ValueError(f"Keyspace '{keyspace}' does not exist. please create it first.")
+
+        self.create_tables()
+
+        # prepare statements
+
+        chunk_counts_cql = f"""
+        SELECT COUNT(*) FROM {keyspace}.chunks
```

Comment:
this won't work in any real system

Reply:
Can you explain why it's won't in any real system? Is it because too many chunks?

Reply:
Removed. I don't need this. Because it's only needed to reduce k when k is less than the total row. K is 10, any meaningful even test data should have more than 10 chunks.